RiotSwiftUI/Modules/Authentication/Registration/Test/UI/AuthenticationRegistrationUITests.swift

```diff
@@ -0,0 +1,144 @@
+//
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import XCTest
+import RiotSwiftUI
+
+@available(iOS 14.0, *)
+class AuthenticationRegistrationUITests: MockScreenTest {
```

Comment:
These tests are useful 👍 . I'd say though that where we are testing for existence of UI elements given certain state, the same test can be expressed in UI-agnostic way, by simply unit testing `ViewState`s with various (`shouldShowPassword` etc) fields. A lot of the times we are really interested in testing some abstract logic (e.g. if username and password are filled in, some action/button becomes enabled etc)

Not suggesting we convert that here, but might be more practical and easier-to-setup way to test the UI

Reply:
I'm not sure if this is useful context: these tests are based on the way the SwiftUI templates were set up.
- [Mock screen states](https://github.com/vector-im/element-ios/blob/develop/RiotSwiftUI/Modules/Template/SimpleUserProfileExample/MockTemplateUserProfileScreenState.swift) for SwiftUI previews
- [UI tests](https://github.com/vector-im/element-ios/blob/develop/RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Test/UI/TemplateUserProfileUITests.swift) based on those screen states

I've been following those patterns and this is the first screen where I notice a lot of reuse of various checks so decided to have multiple checks per state.

Reply:
Yea it makes sense why these were setup and it is a valid way to test the UI. I suppose I was questioning the need to test the UI to begin with, as opposed to testing the pure ViewStates. It seems totally possible to test cases such as "Checks that the username and password text fields are shown along with the next button." without touching UI layer, and it seems to be what you have done in `AuthenticationRegistrationViewModelTests`. I am generally in favour of an approach where the UI becomes so trivial and dumb in its consumption of fully tested view states, that there is very little else to test via UI tests. I am always happy with tests of any kind though 👍 .

Reply:
My only context here is that I can say that these view state tests did catch one error I made in the onboarding flow where I broke something in the view itself so it didn't map to the model correctly. But it would be so much nicer to have snapshot tests for these states instead. Yep, maybe a point for discussion on Friday?